### DISCUSSION TITLE
Lootbox v1.0.0 - Cashback Rewards

### DISCUSSION THREAD

**Discussion by 0x4007** [pavlovcik/uad-ui#601] [Protocol Specifications]
### Discussed in https://github.com/ubiquity/ubiquity-dollar-development/discussions/158

<div type='discussions-op-text'>

<sup>Originally posted by **pavlovcik** May 13, 2022</sup>
# v1.0.0 - Proof of concept - Cashback on ANY NFT Purchases on OpenSea

We want to increase usage of the uAD by giving people a reason to purchase NFTs with uAD. Introducting Ubiquity Credit Cashback Program, similar to credit cards!

+ #158 
+ #165 
+ #166 

## Overview

> アレクサンダー.eth | U B I Q U I T Y:
i came up with a hack to make the opensea thing work

> アレクサンダー.eth | U B I Q U I T Y:
> <img width="1552" alt="Screen Shot 2022-05-13 at 05 26 17" src="https://user-images.githubusercontent.com/4975670/168184188-5c33ed1f-1a11-4595-a17c-3103d5cf6bb0.png">

> アレクサンダー.eth | U B I Q U I T Y:
so basically we modify our linkedin chrome extension (already has great infra, client side!) to modify the opensea ui

> アレクサンダー.eth | U B I Q U I T Y:
then we deploy one or two contracts to middle man the transaction

> アレクサンダー.eth | U B I Q U I T Y:
in these contracts, we take, for example, a 2.5% fee (just like a credit card/payment processor company) and then give them cashback; for example up to 10% in Ubiquity Credits (uCR) or the total of all sum fees (e.g. 2.5% OpenSea, 5% creator fee, 2.5% our processing fee)

> アレクサンダー.eth | U B I Q U I T Y:
this gives us a new revenue source

> アレクサンダー.eth | U B I Q U I T Y:
behind the scenes, we swap the ubiquity dollar for eth/weth used for the purchase and complete the opensea transaction

> T Flex 👀:
i wouldn’t replace the button but add another one

## Smart Contract

The purpose of the "middle-man" contract would be to perform the following:

1. receive the user's uAD
1. set aside uCR valued at 10% of this uAD amount
    1. for security reasons, we should top up this contract manually with uCR
1. check now for the NFT price, and what asset the NFT is priced in (WETH, ETH, DAI, USDC?) and if the uAD amount is valued at 102.5% of the NFT price, not including slippage on the uAD swap.
    1. we should automatically handle choosing and swapping for the correct purchase asset as an extra convenience for the user
1. swap for ETH using the user's uAD (this should be valued at 102.5%+ of the NFT price)
1. purchase NFT with 97.5%, and keep 2.5% of the purchase amount, in the swapped for asset, e.g. ETH
1. send back NFT + uCR "cashback" to user

### Setters

We should be able to set the following variables

1. transaction fee (default at 2.5%)
1. cashback percentage (default at 10%) 
1. account for slippage (default true)

## Remarks

This gives us a reason to do a bond for uAD-ETH and grow its liquidity.</div>

**Comment by 0x4007** [comment]
@sergfeldman idea for lootbox v1.1: add on top of talent marketplaces like [dework.xyz](https://dework.xyz/), [gitcoin.co](https://gitcoin.co/), [layer3.xyz](https://beta.layer3.xyz/)

**Comment by cuddle-dev** [comment]
I am gonna break down this issue into both stuff.

onchain stuff and offchain stuff

In onchain stuff, we will have to implement middle-man contract which basically works 

- [ ] Receive uAD from user, 
- [ ]  Take the fee in uAD
- [ ] Swap uAD with ETH (because most collections only accepts ETH), 
- [ ]  Purchase a nft from seaport contract, sends the nft back to the user

In offchain stuff, we will use genie 3rd-party service to generate calldata and pass it into the middle-man contracts. 
I think we will have to build backend agent(like a relayer) to keep the genie api key hidden.

**Comment by 0x4007** [comment]
I don't recall where we asked for an API key. You mentioned this on our last call. I checked the Discord and the only question I asked was related to documentation. @sergfeldman any info on an API key? 

Otherwise would it be faster to start by using another aggregator? 

Ultimately the bulk of the work should be in the core logic, and what aggregator we connect should basically just be interface work. 

Meaning that while we figure out the API key stuff or any other prerequisite work for another aggregator, perhaps it's wise to start writing the core logic?

**Comment by sergfeldman** [comment]
I don't have specific information about the Genie API key. 
In one of the meetings we recommended Seo to contact the Genie  Discord with technical questions.

When choosing an aggregator, it is worth remembering that gem.xyz belongs to OpenSea
https://opensea.io/blog/announcements/opensea-acquires-gem-to-invest-in-pro-experience/

**Comment by cuddle-dev** [comment]
That's really good news that gem.xyz belongs to OpenSea. It could be another reason why we use gem rather than using genie.

**Comment by cuddle-dev** [comment]
will start handling this stuff!

**Comment by sergfeldman** [comment]
This issue must be divided into several smaller issues.

**Comment by 0x4007** [comment]
> * Lootbox v1.0.0 - Cashback Rewards #158

I agree. It would be helpful if you attempt a draft and I can refine later! My current process is to edit the first comment in this conversation and add a checklist. That way, the GitHub UI allows for quick sub-issue creation with a single button click.

**Comment by hashedMae** [comment]
I can start working on draft contracts for this. Are we going to pull the price in the listed asset via an ABI or would we rather retrieve that on chain?

**Comment by 0x4007** [comment]
I have a couple of reservations. 

1. We should break this issue into separate bounties. We've seen much better results with smaller more tightly scoped bounties.
2. This isn't a strategic priority at the moment. 

If you're particularly motivated to work on this perhaps you can take the lead on breaking down to separate bounties. Otherwise, consider working on the other repos like `bounty-bot` if you want to try something different?
